core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -105,65 +168,255 @@
             } else if (exception[0] instanceof Error) {
                 throw (Error) exception[0];
             } else {
-                throw new RuntimeException(exception[0]);
+                throw new RuntimeException("Run in transaction '"+action+"': "+exception[0].getLocalizedMessage(), exception[0]);
             }
         }
         return result;
     }
 
-    protected static void doCommit(ApplicationContext context, PlatformTransactionManager transactionManager, TransactionStatus transaction) {
+    /**
+     * Safely perform {@link PlatformTransactionManager#commit(TransactionStatus)}.
+     *
+     * This method is responsible for safely logging any notification failures as warnings, rather than failing outright.
+     *
+     * @param context context used to obtain listeners to notify
+     * @param transactionManager manager responsible for execution
+     * @param status transaction status
+     */
+    protected static void doCommit(ApplicationContext context, PlatformTransactionManager transactionManager, TransactionStatus status) {
+        try {
+            fireBeforeCommit(context, transactionManager, status);
+        }
+        catch( Throwable t ){
+            Log.warning(Log.JEEVES, "Commit transaction - before:", t);
+        }
 
-        for (BeforeCommitTransactionListener listener : context.getBeansOfType(BeforeCommitTransactionListener.class).values()) {
-            listener.beforeCommit(transaction);
+        if ( status == null || status.isCompleted()){
+            // not calling return here to preserve previous logic
+            // we can check if the following log messages are seen before taking defensive action
+            Log.debug( Log.JEEVES,"transactionManager.commit called unexpectedly when transaction is already completed ");
         }
 
-        transactionManager.commit(transaction);
+        try {
+            transactionManager.commit(status);
+        }
+        finally {
+            try {
+                fireAfterCommit(context, transactionManager, status);
```

Comment:
In the original code if `transactionManager.commit` fails, the after commit listeners were not executed, with this change seem are executed.

Not very clear if it's relevant to execute them, unless they could handle the transaction commit failure?

Reply:
they can check for the commit failure, made a note in the javadocs and added a log.